src/GasEstimator.ts

```diff
@@ -0,0 +1,234 @@
+import {
+    constants,
+    ContractInteractor,
+    DeployTransactionRequest,
+    ERC20Token,
+    EstimateGasParams,
+    estimateMaxPossibleRelayCallWithLinearFit,
+    RelayTransactionRequest
+} from '@rsksmart/rif-relay-common';
+import { DeployRequest, RelayRequest } from '@rsksmart/rif-relay-contracts';
+import BigNumber from 'bignumber.js';
+import { toWei } from 'web3-utils';
+
+const SUBSIDY = BigNumber(12000);
+
+/**
+ * Estimates the gas consume by relaying a transaction using either a linearFit/standard estimation
+ * @param contractInteractor object containing the contractInteractor
+ * @param request request that contains the relayRequest/deployRequest and metadata
+ * @param relayWorker address of the relayWorker that will execute the transaction
+ * @returns gas estimation from the relayTransaction
+ */
+export const estimateGasRelayTransaction = async (
+    contractInteractor: ContractInteractor,
+    request: RelayTransactionRequest | DeployTransactionRequest,
+    relayWorker: string
+): Promise<BigNumber> => {
+    const { relayRequest, metadata } = request;
+
+    let estimation: BigNumber;
+
+    const tokenEstimation = await estimateGasTokenTransfer(
+        contractInteractor,
+        relayRequest
+    );
+
+    if (checkSignature(metadata.signature)) {
+        estimation = await standardGasEstimation(
+            contractInteractor,
+            request,
+            relayWorker,
+            tokenEstimation
+        );
+    } else {
+        estimation = await linearFitGasEstimation(
+            contractInteractor,
+            relayRequest,
+            tokenEstimation
+        );
+    }
+
+    return estimation;
+};
+
+const checkSignature = (signature: string): boolean => {
+    const bigValue: BigNumber = new BigNumber(signature, 16);
+    if (!bigValue.isZero()) {
+        return true;
+    }
+    return false;
+};
+
+/**
+ * Estimates the gas consume by relaying a transaction using a standard estimation
+ * @param contractInteractor object containing the contractInteractor
+ * @param request request that contains the relayRequest/deployRequest and metadata
+ * @param relayWorker address of the relayWorker that will execute the transaction
+ * @param tokenEstimation gas consume by the token transfer
+ * @returns gas estimation from the relayTransaction
+ */
+export const standardGasEstimation = async (
+    contractInteractor: ContractInteractor,
+    {
+        relayRequest,
+        metadata
+    }: RelayTransactionRequest | DeployTransactionRequest,
+    relayWorker: string,
+    tokenEstimation: BigNumber
+): Promise<BigNumber> => {
+    const { request, relayData } = relayRequest;
+
+    let methodToEstimate;
+    if ('index' in request) {
```

Comment:
We could extract this into a function `isDeployRequest`

Reply:
updated